sdk/nodejs/output.ts

```diff
@@ -445,21 +433,88 @@ export function output<T>(val: Input<T | undefined>): Output<Unwrap<T | undefine
         // 2. That the `isSecret` property is available.
         // 3. That the `.allResources` is available.
         const allResources = getAllResources(val);
-        const newOutput = new Output(
+        const newOutput = new Output<any>(
             val.resources(), val.promise(/*withUnknowns*/ true), val.isKnown, val.isSecret, allResources);
-        return <any>(<any>newOutput).apply(output, /*runWithUnknowns*/ true);
+        return {
+            value: <any>(<any>newOutput).apply(output, /*runWithUnknowns*/ true),
+            isOutput: true,
+         };
     }
     else if (val instanceof Array) {
-        return <any>all(val.map(output));
+        const allValues = [];
+        let hasOutputs = false;
+        for (const v of val) {
+            const ev = outputRec(v);
+
+            allValues.push(ev.value);
+            if (ev.isOutput) {
+                hasOutputs = true;
+            }
+        }
+
+        if (!hasOutputs) {
+            return { value: allValues };
+        }
+
+        const promisedArray = Promise.all(allValues.map(v => getAwaitableValue(v)));
+        const [syncResources, isKnown, isSecret, allResources] = getResourcesAndDetails(allValues);
+        return {
+            value: (new Output<any>(syncResources, promisedArray, isKnown, isSecret, allResources)),
+            isOutput: true,
+        };
     }
     else {
-        const unwrappedObject: any = {};
+        const promisedValues: { key: string; value: any }[] = [];
```

Comment:
I don't exactly follow why we need/want to construct this as an array instead of directly as an object - but I trust there is simplicity somehow in doing this.